storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,756 @@
+//! to determine when to force pending data to be written to `Storage` using the `sync` method. When
+//! calling `close`, all pending data is automatically synced and any open blobs are closed.
+//!
+//! # Pruning
+//!
+//! Old data can be pruned from `Journal` by calling the `prune` method which will remove all blobs
+//! consisting entirely of values older than the given item.
+//!
+//! # Replay
+//!
+//! The `replay` method iterates over multiple blobs concurrently to support fast reading of all
+//! unpruned items into memory.
+
+use super::Error;
+use bytes::BufMut;
+use commonware_runtime::{Blob, Error as RError, Storage};
+use commonware_utils::hex;
+use futures::stream::{self, Stream, StreamExt};
+use prometheus_client::metrics::{counter::Counter, gauge::Gauge};
+use prometheus_client::registry::Registry;
+use std::collections::BTreeMap;
+use std::sync::{Arc, Mutex};
+use tracing::{debug, trace, warn};
+
+/// Configuration for `Journal` storage.
+#[derive(Clone)]
+pub struct Config {
+    /// Registry for metrics.
+    pub registry: Arc<Mutex<Registry>>,
+
+    /// The `commonware-runtime::Storage` partition to use for storing journal blobs.
+    pub partition: String,
+
+    pub items_per_blob: u64,
+}
+
+/// Implementation of `Journal` storage.
+pub struct Journal<B: Blob, E: Storage<B>, const N: usize> {
+    runtime: E,
+    cfg: Config,
+
+    // Blobs are stored in a BTreeMap to ensure they are always iterated in order of their indices.
+    // Indices are consecutive and without gaps.
+    blobs: BTreeMap<u64, B>,
+
+    tracked: Gauge,
+    synced: Counter,
+    pruned: Counter,
+}
+
+impl<B: Blob, E: Storage<B>, const N: usize> Journal<B, E, N> {
+    /// Initialize a new `Journal` instance.
+    ///
+    /// All backing blobs are opened but not read during initialization. The `replay` method can be
+    /// used to iterate over all items in the `Journal`.
+    pub async fn init(runtime: E, cfg: Config) -> Result<Self, Error> {
+        // Iterate over blobs in partition
+        let mut blobs = BTreeMap::new();
+        let stored_blobs = match runtime.scan(&cfg.partition).await {
+            Ok(blobs) => blobs,
+            Err(RError::PartitionMissing(_)) => Vec::new(),
+            Err(err) => return Err(Error::Runtime(err)),
+        };
+        for name in stored_blobs {
+            let blob = runtime
+                .open(&cfg.partition, &name)
+                .await
+                .map_err(Error::Runtime)?;
+            let index = match name.try_into() {
+                Ok(index) => u64::from_be_bytes(index),
+                Err(nm) => return Err(Error::InvalidBlobName(hex(&nm))),
+            };
+            debug!(blob = index, "loaded blob");
+            blobs.insert(index, blob);
+        }
+        if !blobs.is_empty() {
+            // Check that there are no gaps in the blob numbering, which would indicate missing data.
+            let mut it = blobs.iter();
```

Comment:
I believe you can just do a `keys()` iter here (to avoid the `.0`)?

Reply:
done